src/host/HostShareNUMA.cc

```diff
@@ -1035,37 +1046,223 @@ bool HostShareNUMA::schedule_nodes(NUMANodeRequest &nr, unsigned int threads,
+    // -------------------------------------------------------------------------
+    unsigned long n_hp = 0;
+
+    HostShareNode::HugePage * hpage = nullptr;
+
+    if (hpsz_kb != 0)
+    {
+        auto jt = node->pages.find(hpsz_kb);
+
+        if ( jt == node->pages.end())
+        {
+            return -1; // Node has no huge pages of requested size
+        }
+
+        hpage = &(jt->second);
+
+        n_hp  = sr.mem / hpsz_kb; //sr.mem = SUM(sr.nodes.memory)
+        long long n_fhp = hpage->nr - hpage->usage;
+
+        if (n_fhp <= 0 || n_fhp < n_hp)
+        {
+            return -1; //Node has not enough free huge pages (n_fhp)
+        }
+    }
+
+    if (!do_alloc)
+    {
+        return 0;
+    }
+
+    // -------------------------------------------------------------------------
+    // Save CPUs and memory allocation to the VM NUMA node
+    // -------------------------------------------------------------------------
+    std::string cpu_ids;
+
+    node->ls_cpus(false, cpu_ids);
+
+    node->mem_usage += sr.mem;
+
+    if (hpsz_kb != 0 && hpage != nullptr)
+    {
+        hpage->usage += n_hp;
+
+        node->update_hugepages();
+    }
+
+    node->update_memory();
+
+    for (auto &vm_node : sr.nodes)
+    {
+        vm_node->replace("NODE_ID", node_id);
+        vm_node->replace("MEMORY_NODE_ID", node_id);
+        vm_node->replace("CPUS", cpu_ids);
+    }
+
+    return 0;
+}
+
+/* -------------------------------------------------------------------------- */
+/* -------------------------------------------------------------------------- */
+
+int HostShareNUMA::make_topology(HostShareCapacity &sr, int vm_id, bool do_alloc)
+{
+    // -------------------------------------------------------------------------
+    // Regular VMS (not pinned and no NUMA affinity)
+    // -------------------------------------------------------------------------
     if ( sr.topology == 0 || sr.nodes.empty() )
     {
         return 0;
     }
 
-    sr.topology->vector_value("THREADS", v_t);
-    sr.topology->vector_value("CORES", c_t);
-    sr.topology->vector_value("SOCKETS", s_t);
+    // -------------------------------------------------------------------------
+    // NUMA node affinity
+    // -------------------------------------------------------------------------
+    int affinity = -1;
+    unsigned long hpsz_kb = 0;
+
+    std::string policy      = sr.topology->vector_value("PIN_POLICY");
+    HostShare::PinPolicy pp = HostShare::str_to_pin_policy(policy);
+
+    sr.topology->vector_value("NODE_AFFINITY", affinity);
 
     sr.topology->vector_value("HUGEPAGE_SIZE", hpsz_kb);
     hpsz_kb = hpsz_kb * 1024;
 
-    std::string policy      = sr.topology->vector_value("PIN_POLICY");
-    HostShare::PinPolicy pp = HostShare::str_to_pin_policy(policy);
+    if (affinity != -1)
+    {
+        return make_affined_topology(sr, affinity, hpsz_kb, do_alloc);
+    }
+
+    if (hpsz_kb != 0 && pp == HostShare::PP_NONE)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

We observed that you are using unintended operations or assignments that could lead to logical errors, vulnerabilities, or undefined behavior. To prevent these issues, ensure that you follow best practices, such as using explicit comparisons and clear operator usage, and using parentheses to clarify intent. Learn more - https://cvefeed.io/cwe/detail/cwe-480-use-of-incorrect-operator